### PR TITLE
Issue #3047: Add setting to add phone number field to contact forms

### LINKED ISSUE #3047: [UX] Add Phone field to Contact forms
The current proposal is to add a few additional fields to contact forms, with settings for them:

1) markup or message field, like a body (setting: textarea field - default to empty)
2) phone number field (setting: on/off checkbox - default to off)

# Original issue

I'd love to see the Contact module/form made fieldable. I know that Webform exists and is often used for this purpose, but it does a lot of extra stuff that is unnecessary when you simply want to add a Phone field to the existing Contact form (for example).

I envisage something like this:

- Fields can be added in the same way as nodes
- Fields can be re-ordered
- Contrib modules (like Captcha) can add their fields which makes them re-orderable too
- Contact settings (like recipients/categories) can be managed on a form like the Create Content Type form
- Maybe go further and make Contact an entity, and then 'categories' become bundles (e.g. you can add different fields for different departments; so that when someone contacts your support department, they can also enter their customer ID)

### PR DESCRIPTION
**Pull request by ghost**
Fixes https://github.com/backdrop/backdrop-issues/issues/3047

Replaces https://github.com/backdrop/backdrop/pull/3374

### CODE REVIEW

File: core/modules/contact/contact.pages.inc [hunk 1]
```diff
@@ -76,6 +76,14 @@ function contact_site_form($form, &$form_state) {
     '#default_value' => $user->uid ? $user->mail : '',
     '#required' => TRUE,
   );
+  if (!empty($config->get('phone'))) {
+    $form['phone'] = array(
+      '#type' => 'textfield',
```
Comment (indigoxela) [inline core/modules/contact/contact.pages.inc]: Have you considered to use [input of type "tel"](https://github.com/backdrop/backdrop/blob/10167aae01ea19486e532871cfbaf238adebfa14/core/modules/system/system.module#L293) instead? Mobile browsers might benefit.

Comment (ghost) [reply core/modules/contact/contact.pages.inc]: Does it allow for different characters, like '+', '()', etc.?

E.g.: `+61 400 123 456` or `(02) 6654 9870`

Comment (indigoxela) [reply core/modules/contact/contact.pages.inc]: It doesn't restrict anything, as long as there's no "pattern" attribute - actually identical to type "text".

Here's an [example (with pattern, though)](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/tel)

This is just a suggestion. We have it in core, so why not using it. :wink:

Comment (ghost) [reply core/modules/contact/contact.pages.inc]: Done, thanks!

File: core/modules/contact/contact.pages.inc [hunk 2]
```diff
@@ -219,6 +226,14 @@ function contact_personal_form($form, &$form_state, $recipient) {
...
```
Comment (indigoxela) [inline core/modules/contact/contact.pages.inc]: This one (personal contact form) is still "textfield". :wink:

File: core/modules/contact/contact.pages.inc [hunk 3]
```diff
@@ -76,6 +76,13 @@ function contact_site_form($form, &$form_state) {
     '#default_value' => $user->uid ? $user->mail : '',
     '#required' => TRUE,
   );
+  if (!empty($config->get('phone'))) {
```
Comment (indigoxela) [inline core/modules/contact/contact.pages.inc]: These are the killers. See https://joomla.stackexchange.com/questions/5565/fatal-error-cant-use-method-return-value-in-write-context

Comment (ghost) [reply core/modules/contact/contact.pages.inc]: Thanks @indigoxela & @docwilmot!

File: core/modules/contact/contact.admin.inc [hunk 4]
```diff
@@ -22,6 +22,17 @@ function contact_settings_form($form, &$form_state) {
...
   );
+  $form['fields']['phone'] = array(
+    '#type' => 'select',
+    '#title' => t('Phone number'),
+    '#description' => t('Allow users to enter their phone number.'),
+    '#options' => array(
+      'optional' => t('Optional'),
+      'required' => t('Required'),
+    ),
+    '#empty_value' => '',
```
Comment (klonos) [inline core/modules/contact/contact.admin.inc:33]: Can we please change this one to:

```suggestion
    '#empty_value' => t('Disabled'),
```

...I believe that it makes more sense in this context, as compared to the default `- None -`

Comment (ghost) [reply core/modules/contact/contact.admin.inc:33]: I disagree. 'None' means there's no field (which is true). 'Disabled' implies that the field is there but cannot be edited (which is not true).

Comment (klonos) [reply core/modules/contact/contact.admin.inc:33]: Fair enough re "disabled". ...how about "hidden" then? ...was actually going for consistency and was thinking of this:

image

Comment (ghost) [reply core/modules/contact/contact.admin.inc:33]: I don't know... 'Hidden' implies the field is added but not visible. I really don't see the problem with 'None'... Maybe bring it up in the issue and see what others think.

Comment (docwilmot) [reply core/modules/contact/contact.admin.inc:33]: I believe "none" makes more sense.

File: core/modules/contact/contact.admin.inc [hunk 5]
```diff
@@ -22,6 +22,17 @@ function contact_settings_form($form, &$form_state) {
...
+    '#default_value' => $config->get('phone'),
+  );
   $form['fields']['copy'] = array(
```
Comment (docwilmot) [inline core/modules/contact/contact.admin.inc:36]: Perhaps this is not within the scope of this issue but I don't get why this fieldset **was** called fields before this PR. It actually makes sense now that we have the phone field in there, but now I don't think the "copy" checkbox fits.

Comment (ghost) [reply core/modules/contact/contact.admin.inc:36]: The 'copy' checkbox *is* actually a field that appears on the Contact form though:

image

Comment (docwilmot) [reply core/modules/contact/contact.admin.inc:36]: Ah, truth. Ignore me.

**Comment by ghost** [comment]
Merged into 1.x.

**Pull request merged**